sympy/matrices/expressions/matmul.py

```diff
@@ -44,14 +46,17 @@ def __new__(cls, *args, **kwargs):
         args = filter(lambda i: cls.identity != i, args)
         args = list(map(sympify, args))
         obj = Basic.__new__(cls, *args)
-        factor, matrices = obj.as_coeff_matrices()
+
+        _, others = obj._as_commutative_scalars_others()
+        matrices, _ = sift(others, lambda x: x.is_Matrix, binary=True)
```

Comment:
This logic is incorrect. It will require the shapes to match "across" noncommutatives. Consider

```py
n, m, k, j = symbols('n m j k', integer=True)
A = MatrixSymbol('A', n, m)
B = MatrixSymbol('B', m, k)
C = MatrixSymbol('C', k, j)
BB = symbols('BB', commutative=False)
```

```py
>>> (A*B*C).subs(B, BB)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./sympy/core/basic.py", line 1002, in subs
    rv = rv._subs(old, new, **kwargs)
  File "./sympy/core/basic.py", line 1116, in _subs
    rv = fallback(self, old, new)
  File "./sympy/core/basic.py", line 1093, in fallback
    rv = self.func(*args)
  File "./sympy/matrices/expressions/matmul.py", line 54, in __new__
    validate(*matrices)
  File "./sympy/matrices/expressions/matmul.py", line 224, in validate
    raise ShapeError("Matrices %s and %s are not aligned"%(A, B))
sympy.matrices.common.ShapeError: Matrices A and C are not aligned
```

Reply:
I'm not sure about how the resulting shape of a multiplication between a matrix and a noncommutative scalar can be

Reply:
This is part of the reason it's a NotImplementedError right now :)

We could ignore noncommutatives for shape checks, but the problem is if the leftmost or rightmost matrix is a noncommutative, then the MatMul cannot have a meaningful shape. I really don't know if there is a way to make this work. The way that shape checking is done in the matrix expressions, it's impossible to have something that's shape invalid, even as an intermediate expression.

Reply:
It sounds like the shape checking needs to be relaxed so that it only raises when the shapes are known to be inconsistent. Then this should be fine because n and m *might* be equal:
```julia
In [18]: A = MatrixSymbol('A', 2, n)                                                                                              

In [19]: B = MatrixSymbol('B', m, 3)                                                                                              

In [20]: A*B                                                                                                                      
---------------------------------------------------------------------------
ShapeError
```
There are harder cases though like `(2 x n)*(2 x n)*(3 x 4)` where the first product only works if `n==2` and the second only works if `n==3`. It might be hard to detect that so an error might not get raised but it's more important to allow correct operations than disallow incorrect ones.

Actually it's not that hard to detect it. We just have to check `And(Eq(n, 2), Eq(n, 3))` which simplifies to `False`.

Reply:
Is a noncommutative "scalar" something that can possibly be a matrix?

Reply:
I would sum up the ideas to 
1. Allow to build matrix multiplication expression with noncommutative symbols as the original post wants.
2. Querying for the matrix shape may raise `NotImplementedError` or may give an another indeterminate symbolic object if the matrix multiplication has a noncommutative scalar.
3. However, matrix shape checking should be relaxed to allow some indeterminates, but only restricts some unsatisfiable condition.

`(2 x n)*(2 x n)*(3 x 4)` case can easily be handled by testing the truth value of 
`Eq(n, 2) & Eq(n, 3)`. If the boolean expression correctly simplifies, it should give `False`.

Reply:
I've also thought this before. There are advantages to having it consider n and m unequal for shape testing purposes, though. 

Perhaps a better way is to only do type checking on evaluation. So

```py
A = MatrixSymbol('A', n, n)
B = MatrixSymbol('A', m, m)
MatMul(A, B)
```

would not raise any exceptions. But 

```py
MatMul(A, B).doit()
``` 

would. `A*B` calls creates a MatMul and calls doit() so it would as well. I think this fits with the general design of matrix expressions where nothing evaluates until you call doit().

Then things that want to substitute a noncommutatives and then replace them at the end can. subs would need to be made to not evaluate (which could be annoying for other things). I'm not sure about the use-case of the original issue. 

Related to this, we also had a discussion in https://github.com/sympy/sympy/pull/15948 about how to handle `MatMul()` and `MatAdd()` with no arguments. The first approach was to make a shape that always compares equal. The approach we ended up using was to create a GenericIdentity and GenericZeroMatrix that don't do shape comparisons. However, these really only work because we can remove them aggressively in the constructors, even without calling doit(), so it's not really the greatest thing IMO (e.g., `MatMul(MatAdd(), MatrixSymbol("A", n, n))` gives an error).

Reply:
Even if we do go the route of making n and m compare equal, if we do any more advanced checking like the Eq checks, we should move that to doit(). The general case of checking if a system of Diophantine equations is satisfiable is enough computation that it shouldn't be done in the constructor.

Reply:
What does `MatMul(A, B).doit()` do if `A` and `B` are matrix symbols?

Reply:
Right now just `MatMul(A, B)` gives a ShapeError exception if the shapes don't match. If the shapes do match then doit() does nothing because the expression cannot be evaluated further.